sklearn/decomposition/ksvd.py

```diff
@@ -0,0 +1,343 @@
+"""K-SVD dictionary learning algorithm.
+"""
+
+
+import numpy
+import scipy.sparse.linalg
+
+from ..base import BaseEstimator
+from .dict_learning import sparse_encode, SparseCodingMixin
+from ..preprocessing import normalize
+from ..utils import check_array, check_random_state
+from ..utils.extmath import norm
+from ..utils.random import choice
+
+
+# atoms with norm below are considered zero
+_ATOM_NORM_TOLERANCE = 1e-6
+
+
+def _worst_represented_example(X, dictionary, sparse_codes):
+    """Find the sample that has the most representation error."""
+
+    residuals = X - numpy.dot(sparse_codes, dictionary)
+    errors_squared = numpy.sum(residuals * residuals, axis=1)
+    worst_index = numpy.argmax(errors_squared)
+    return X[worst_index, :]
+
+
+def _decompose_svd(residual):
+    """Decompose the residual using 1-rank SVD."""
+
+    # Sparse version is used because non-sparse does not allow to
+    # set the desired number of singular values.
+    U, S, V = scipy.sparse.linalg.svds(residual, 1)
+    # U is unitary, so d is normalized
+    d = U.ravel()
+    g = numpy.dot(V.T, S).ravel()
+    return d, g
+
+
+def _decompose_projections(residual, g_old):
+    """Decompose the residual using SVD-approximating projections."""
+
+    d = numpy.dot(residual, g_old)
+    d_norm = norm(d)
+    if d_norm >= _ATOM_NORM_TOLERANCE:
+        d /= d_norm
+    g = numpy.dot(residual.T, d)
+    return d, g
+
+
+def _atom_update(X, atom_index, dictionary, sparse_codes, approximate_svd):
+    """Update single dictionary atom and the corresponding codes."""
+
+    atom_usages = numpy.nonzero(sparse_codes[:, atom_index])[0]
+    if len(atom_usages) == 0:
+        atom = _worst_represented_example(X, dictionary, sparse_codes)
+        dictionary[atom_index, :] = atom / norm(atom)
+    elif len(atom_usages) == 1:
+        # update is trivial if the atom is used in only one sample
+        atom = X[atom_usages[0], :]
+        atom_norm = norm(atom)
+        dictionary[atom_index, :] = atom / atom_norm
+        sparse_codes[atom_usages[0], atom_index] = atom_norm
+    else:
+        dictionary[atom_index, :] = 0
+        representation = numpy.dot(sparse_codes[atom_usages, :], dictionary)
+        residual = (X[atom_usages, :] - representation).T
+        if approximate_svd:
+            atom, atom_codes = _decompose_projections(
+                residual, sparse_codes[atom_usages, atom_index])
+        else:
+            atom, atom_codes = _decompose_svd(residual)
+        dictionary[atom_index, :] = atom
+        sparse_codes[atom_usages, atom_index] = atom_codes.T
+
+    return dictionary, sparse_codes
+
+
+def _reconstruction_error(X, dictionary, sparse_codes):
+    """Calculate the total reconstruction error for data."""
+    errors = X - numpy.dot(sparse_codes, dictionary)
+    return norm(errors)
+
+
+def _init_dictionary_from_samples(X, n_components, random_state):
```

Comment:
In `dict_learning`, we use an svd for finding initial atoms. I think it would be more consistent to allow doing it in k-SVD. Maybe provide two initialization  methods in `dict_learning` module ?